mergin/report.py

```diff
@@ -0,0 +1,237 @@
+# in geodiff lib (MIT licence)
+# ideally there should be pygeodiff api for it
+def parse_gpkgb_header_size(gpkg_wkb):
+    """ Parse header of geopackage wkb and return its size """
+    # some constants
+    no_envelope_header_size = 8
+    flag_byte_pos = 3
+    envelope_size_mask = 14
+
+    try:
+        flag_byte = gpkg_wkb[flag_byte_pos]
+    except IndexError:
+        return -1  # probably some invalid input
+    envelope_byte = (flag_byte & envelope_size_mask) >> 1
+    envelope_size = 0
+
+    if envelope_byte == 1:
+        envelope_size = 32
+    elif envelope_byte == 2:
+        envelope_size = 48
+    elif envelope_byte == 3:
+        envelope_size = 48
+    elif envelope_byte == 4:
+        envelope_size = 64
+
+    return no_envelope_header_size + envelope_size
+
+
+class ChangesetReportEntry:
+    """ Derivative of geodiff ChangesetEntry suitable for further processing/reporting """
+    def __init__(self, changeset_entry, geom_idx, geom):
+        self.table = changeset_entry.table.name
+        self.geom_type = geom["type"]
+        self.crs = geom["srs_id"]
+
+        if changeset_entry.operation == changeset_entry.OP_DELETE:
+            self.operation = "delete"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = None
+        elif changeset_entry.operation == changeset_entry.OP_UPDATE:
+            self.operation = "update"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = changeset_entry.new_values[geom_idx]
+        elif changeset_entry.operation == changeset_entry.OP_INSERT:
+            self.operation = "insert"
+            self.old_geom = None
+            self.new_geom = changeset_entry.new_values[geom_idx]
+
+        self.count = None
+        self.length = None
+        self.area = None
+
+        if self.geom_type == "LINESTRING":
+            # we calculate change in length, for attributes changes only we set to 0
+            self.metric = "length"
+            if self.operation == "delete":
+                self.length = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.length = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.length = self.measure(self.new_geom)
+        elif self.geom_type == "POLYGON":
+            # we calculate change in area, for attributes changes only we set to 0
+            self.metric = "area"
+            if self.operation == "delete":
+                self.area = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.area = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.area = self.measure(self.new_geom)
+        else:
+            # regardless of geometry change count as 1
+            self.metric = "count"
+            self.count = 1
```

Comment:
do I understand correctly, that if I have a linestring, I will only get length reported, but not count?

Reply:
yes, but maybe it makes sense to have both